### PR TITLE
Add SOURCE_DATE_EPOCH support when building packages

### PR DESCRIPTION
**Pull request by Secrus**
Resolves: python-poetry/poetry#3531

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code. (Not sure where the docs about that should go, since poetry-core doesn't have separate docs)

### CODE REVIEW

**Review by trag1c — CHANGES_REQUESTED** [review]
<sub>i accidentally submitted a review in the wrong tab</sub>

**Review by radoering — CHANGES_REQUESTED** [review]

File: src/poetry/core/masonry/builders/sdist.py [hunk 1]
```diff
@@ -93,26 +93,25 @@ def build(
...
-            tar.addfile(tar_info, BytesIO(pkg_info))
+            self.add_file_to_tar(tar, pjoin(tar_dir, "PKG-INFO"), pkg_info)
         finally:
             tar.close()
             gz.close()
 
         logger.info(f"Built <comment>{target.name}</comment>")
         return target
 
+    def add_file_to_tar(
```
[the remark this answers is not in the document]
Comment (Secrus) [reply src/poetry/core/masonry/builders/sdist.py:107]: Honestly, I didn't understand the reason for `clean_tarinfo` to be a `classmethod`, so I don't think I will be changing those unless you see any of those decisions being wrong.

File: tests/masonry/builders/test_sdist.py [hunk 2]
```diff
@@ -514,26 +515,6 @@ def get_ignored_files(self, folder: Path | None = None) -> list[str]:
...
-        # all last modified times should be set to a valid timestamp
-        for tarinfo in tar.getmembers():
-            if tarinfo.name in [
-                "my_package-1.2.3/setup.py",
-                "my_package-1.2.3/PKG-INFO",
-            ]:
-                # generated files have timestamp set to 0
-                assert tarinfo.mtime == 0
-                continue
-            assert tarinfo.mtime > 0
```
[the remark this answers is not in the document]
Comment (Secrus) [reply tests/masonry/builders/test_sdist.py]: Overzealous deletion from my side, restored

File: src/poetry/core/masonry/builders/sdist.py [hunk 3]
```diff
@@ -416,6 +414,20 @@ def clean_tarinfo(cls, tar_info: TarInfo) -> TarInfo:
...
 
         return ti
+
+    @staticmethod
+    def _get_archive_mtime() -> int:
+        if source_date_epoch := os.getenv("SOURCE_DATE_EPOCH"):
+            try:
+                return int(source_date_epoch)
+            except ValueError:
+                logger.info(
```
[the remark this answers is not in the document]
Comment (Secrus) [reply src/poetry/core/masonry/builders/sdist.py]: Redid the rules, let me know what you think.

File: src/poetry/core/masonry/builders/wheel.py [hunk 4]
```diff
@@ -491,6 +491,30 @@ def _write_to_zip(
...
+            # The default is a fixed timestamp rather than the current time, so
+            # that building a wheel twice on the same computer can automatically
+            # give you the exact same result.
+            logger.info(
+                "SOURCE_DATE_EPOCH environment variable not set or value is not an int, setting zipinfo date to default=%s",
+                default,
+            )
+            return default
+        else:
+            if _env_date[0] < 1980:
```
[the remark this answers is not in the document]
Comment (Secrus) [reply src/poetry/core/masonry/builders/wheel.py:520]: Done

File: tests/masonry/builders/test_sdist.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
 
 
 if TYPE_CHECKING:
+    from _pytest.logging import LogCaptureFixture
+    from _pytest.monkeypatch import MonkeyPatch
```
Comment (radoering) [inline tests/masonry/builders/test_sdist.py:29]: cf #772 😉

LGTM apart from that and a conflict to resolve.

**Review by radoering — APPROVED** [review]

**Pull request merged**